### ISSUE TITLE
Use `elastic-package benchmark stream` outside package directory

### ISSUE THREAD

**Issue by ruflin** [elastic/elastic-package#1640]
To make the usage of `elastic-package benchmark stream` as easy as possible for others to use, ideally the check out of the integrations source code repository is not needed. Instead, a user should be able to specify the package(s) they want to ingest data for. This could look similar to to ingest mysql and nginx data:

``` 
elastic-package benchmark stream --packages=nginx,mysql
```

To be able to run this command outside integrations repository directory, I can see two approaches:

* Source packages are available: https://github.com/elastic/elastic-package/issues/1577
* Load the integrations repo in the background

https://github.com/elastic/elastic-package/issues/1577 will likely take longer to implement this is why I would also like to discuss option 2. `elastic-package` uses git to check out the integrations repo in `.elastic-package` (and keep it up to date). It would then load the template files from there. It means, all stream commands would always be run from the "main" branch of the integrations repo. A dependency on "git" is added but not sure if this would be a big issue.

Side note: One odd thing I noticed writing the above example is that the benchmark prefix is in there. It makes sense from a historical perspective but for a user wanting to just ingest data, it might be surprising.

**Comment by ruflin** [comment]
@aspacca @jsoriano Would be great to get your take on the above, especially if option 2 is an option.

**Comment by jsoriano** [comment]
On a first thought this looks like something we could also implement as an external tool, or as a CI pipeline that can be executed on demand with the packages as parameters, what might also help to have a history of executions.

Adding a dependency on git is not an issue, we already import a git client (and a github one). What I see more problematic is to tie this to the integrations repository, this is a dependecy we don't have now, and there can be packages in other repositories.

I also wonder if this is something that we would like to have in other commands, it hasn't been so problematic so far to clone the repository with the packages for other commands. We could also have `elastic-package test --packages=nginx,mysql` or `elastic-build --packages=nginx,mysql`. What makes benchmarks different?

If we decide that this is actually needed, I wouldn't tie it to the integrations repository, and I would consider reviewing other commands in case this applies to more cases. To don't tie it to any repository we could have an additional flag:
```
elastic-package benchmark stream --repository=https://github.com/elastic/integrations.git --packages=nginx,mysql
```

In the mid-long term I would like to see https://github.com/elastic/elastic-package/issues/1577 implemented and possibly publishing the source packages in the package registry, so we continue centralizing discoverability of packages there.

**Comment by aspacca** [comment]
> I also wonder if this is something that we would like to have in other commands, it hasn't been so problematic so far to clone the repository with the packages for other commands. We could also have `elastic-package test --packages=nginx,mysql` or `elastic-build --packages=nginx,mysql`. What makes benchmarks different?

I don't think there should be any difference. Could you share a command execution example/link to code where the repository with the packages is cloned?
I agree we should just rely on what we already have and in case just align on the command flag/usage (referring to the example you made with `--repository=https://github.com/elastic/integrations.git`)


> On a first thought this looks like something we could also implement as an external tool, or as a CI pipeline that can be executed on demand with the packages as parameters, what might also help to have a history of executions.

I'm not sure I exactly get this :)
I don't see an external tool and a CI pipeline to be an implementation alternative to each other of what we want to achieve.
Specifically I would exclude recurring to a CI pipeline: the stream command is specifically intended as a developer tool that you want to run with as much flexibility as you can, it does not apply to the context of CI :)

An external tool just like a bash script that "translates" something like `run_benchmark_stream.sh --packages=nginx,mysql` to multiple `elastic-package benchmark stream` executions from the "nginx" and "mysql" package roots, taking care of cloning the repo as well, will indeed achieve the same goal, and I see it faster and simpler to implement than doing the equivalent from `elastic-package`.

@ruflin I would consider the above temporary solution instead, while we wait for source packages availability, and only after that add the support directly in `elastic-package` as you proposed.

**Comment by ruflin** [comment]
> also implement as an external tool / I would consider the above temporary solution instead,

I would like to stay away from an additional tool especially for the stream command. Users should be able to just run it. Otherwise we have users now start using and external script (with limited functionality) and later on we have to ask them to migrate to elastic-package.

> other commands

Agree, would be great if would work for other commands too

> integrations repo

It should work with any repo that has the right structure. At the same time, for a user that just got started, the integrations repo should be the default or users should be able to set it as default.

**Comment by jsoriano** [comment]
> Side note: One odd thing I noticed writing the above example is that the benchmark prefix is in there. It makes sense from a historical perspective but for a user wanting to just ingest data, it might be surprising.

Yep, I think this was confusing me, I was relating this to benchmarks all the time.

Is then the idea to use this stream subcommand to generate data for any use case? Also for users just wanting for example to see data in dashboards?

Or if this is only for developers, on what use cases a developer would use this command without having a clone of the source of the package?

**Comment by aspacca** [comment]
> > Side note: One odd thing I noticed writing the above example is that the benchmark prefix is in there. It makes sense from a historical perspective but for a user wanting to just ingest data, it might be surprising.
> Yep, I think this was confusing me, I was relating this to benchmarks all the time.

indeed having `stream` as subcommand of `benchmark` is far from ideal, as @ruflin mentioned and as your confusion confirms :)

happy to move it around to a separated command


> Is then the idea to use this stream subcommand to generate data for any use case? Also for users just wanting for example to see data in dashboards?

yes, that's one of the case

> Or if this is only for developers, on what use cases a developer would use this command without having a clone of the source of the package?

it is not only for developers, but there could be the use case of a developer wanting to look at the impact of a stack release to the latest version of an integration (dashboards, whatever). it is probably the case that the developer has already a clone of the source of the package, so we are talking about a condition that will happen quite rarely, but ideally running `stack up` and `benchmark stream` for that use case could not require to have the repo cloned

I'd like to surface another use case
a developer working on `elastic-package`: at least for my experience with the `benchmark` subcommand, having to move in the terminal between the elastic-package repo and the integrations repo is a minor annoyance :)
